### PR TITLE
rmi and build cache miss performance improvements

### LINKED ISSUE #16771: Terrible `docker rmi` performance
Currently, a daemon with >50 images takes a ridiculous amount of time to honor a single `rmi` request.

After investigation, the root cause of the issue is [this line](https://github.com/docker/docker/blob/ec1d8d7f2a8f563deb61df2b35aa8a9d6a07c3e2/daemon/image_delete.go#L282). The `HasChildren` function is incredibly complex: it calls into [`graph.ByParent()`](https://github.com/docker/docker/blob/ec1d8d7f2a8f563deb61df2b35aa8a9d6a07c3e2/graph/graph.go#L391) that walks the entire graph just for the sake of determining if an image has children.

Having a "fast-exit" (e.g., stop the loop when a first children is found) won't solve the most common case. We should look into maybe removing that `ByParent()` function, and maintain a reference count for each graph entry such as delete would fail immediately unless the counter is down to 0.

I would love to fix this for 1.9.0 but still haven't found the time to do it myself :-( Ping @LK4D4 @runcom if interested!

PS: the issue is easily reproduced by doing a `docker pull --all-tags` on `ubuntu` or `debian`.

### PR DESCRIPTION
**Pull request by runcom**
This patch also adds integration tests for images and rmi commands that weren't covered anywhere AFIK

Fixes #16771.

### CODE REVIEW

**Comment by runcom** [comment]
ping @icecrime @LK4D4

**Comment by LK4D4** [comment]
@runcom need rebase btw :)

**Comment by runcom** [comment]
@LK4D4 rebased

File: graph/graph.go [hunk 1]
```diff
@@ -424,11 +462,10 @@ func (graph *Graph) Release(sessionID string, layerIDs ...string) {
 // A head is an image which is not the parent of another image in the graph.
 func (graph *Graph) Heads() map[string]*image.Image {
 	heads := make(map[string]*image.Image)
-	byParent := graph.ByParent()
 	graph.walkAll(func(image *image.Image) {
 		// If it's not in the byParent lookup table, then
```
Comment (coolljt0725) [inline graph/graph.go]: Maybe `If it's not in the byParent lookup table` this comment should be removed? since there is 
no `byParent` lookup table  :)

Comment (runcom) [inline graph/graph.go]: yup thanks @coolljt0725 :)

**Comment by runcom** [comment]
I've also added https://github.com/runcom/docker/commit/84ca81ca4b0ff399fd5bf545043f5c6ba56632b8 which is fixing `graph/_tmp` leak

File: daemon/daemon.go [hunk 2]
```diff
@@ -1142,6 +1142,14 @@ func (daemon *Daemon) GetRemappedUIDGID() (int, int) {
...
 // returned if the parent image cannot be found.
 func (daemon *Daemon) ImageGetCached(imgID string, config *runconfig.Config) (*image.Image, error) {
+	// for now just exit if imgID has no children.
+	// maybe parentRefs in graph could be used to store
+	// the Image obj children for faster lookup below but this can
+	// be quite memory hungry.
+	if !daemon.Graph().HasChildren(imgID) {
+		return nil, nil
+	}
+
```
Comment (aaronlehmann) [inline daemon/daemon.go]: `HasChildren` calls `ByParent` internally. `ByParent` returns a map indexed by the parent ID, which is similar to what this function constructs below. Couldn't this function call `ByParent`, check `len(parents[imgID])`, and if it's nonzero, use the map returned by `ByParent` to find the earliest created child? That would avoid an extra walk of the graph.

Comment (aaronlehmann) [inline daemon/daemon.go]: Sorry, missed your change to `HasChildren`, as it's later in the diff. This makes more sense now. I still think using `ByParent` would make more sense than `Map` in this function, but it's unrelated to your change.

Comment (runcom) [inline daemon/daemon.go]: @aaronlehmann btw I think you're right and we could remove this whole for:
```
imageMap := make(map[string]map[string]struct{})
 	for _, img := range images {
 		if _, exists := imageMap[img.Parent]; !exists {
 			imageMap[img.Parent] = make(map[string]struct{})
 		}
 		imageMap[img.Parent][img.ID] = struct{}{}
 	}
```
and assign the map from `ByParent` to `images` (some lines below), am I correct?

Comment (runcom) [inline daemon/daemon.go]: so there would be only 1 loop to get the parent->children tree (in `ByParent`)

File: graph/graph.go [hunk 3]
```diff
@@ -119,12 +120,13 @@ func NewGraph(root string, driver graphdriver.Driver, uidMaps, gidMaps []idtools
...
-		retained: &retainedLayers{layerHolders: make(map[string]map[string]struct{})},
-		uidMaps:  uidMaps,
-		gidMaps:  gidMaps,
+		root:       abspath,
+		idIndex:    truncindex.NewTruncIndex([]string{}),
+		driver:     driver,
+		retained:   &retainedLayers{layerHolders: make(map[string]map[string]struct{})},
+		uidMaps:    uidMaps,
+		gidMaps:    gidMaps,
+		parentRefs: make(map[string]int),
```
Comment (aaronlehmann) [inline graph/graph.go]: Please add a comment explaining that accesses to this map need to be protected by `graph.imageMutex.Lock` on the key.

File: daemon/daemon.go (hunk 2 above)
Comment (aaronlehmann) [inline daemon/daemon.go]: Yes, that's what I had in mind. `ByParent` returns a flat list of children instead of a nested map, but it looks like it doesn't matter since `ImageGetCached` just loops over the map keys in the nested map it creates.

Comment (runcom) [inline daemon/daemon.go]: no ok, I know had a look at the code :)

File: graph/graph.go (hunk 3 above)
Comment (runcom) [inline graph/graph.go]: :+1:

File: graph/graph.go [hunk 4]
```diff
@@ -292,6 +308,11 @@ func (graph *Graph) Register(img *image.Image, layerData io.Reader) (err error)
 		return err
 	}
 	graph.idIndex.Add(img.ID)
+
+	graph.imageMutex.Lock(img.Parent)
+	graph.parentRefs[img.Parent]++
+	graph.imageMutex.Unlock(img.Parent)
+
```
Comment (aaronlehmann) [inline graph/graph.go]: I think there is a possible deadlock here given that `attemptIDReuse` (see https://github.com/docker/docker/pull/16952/files#diff-7e2a216082b6bd557c7ecebd68305e2eR576) may lock images in a different order.

cc @tonistiigi

Comment (tonistiigi) [inline graph/graph.go]: I'm not sure I get how this is a deadlock if you only lock one image at a time.

Comment (aaronlehmann) [inline graph/graph.go]: `Register` holds a second lock: https://github.com/docker/docker/pull/16890/files#diff-6291e801306012e55f373a86213ee4d5L244

File: integration-cli/docker_cli_images_test.go [hunk 5]
```diff
@@ -197,3 +196,42 @@ func (s *DockerSuite) TestImagesEnsureDanglingImageOnlyListedOnce(c *check.C) {
...
+	dockerfile := `
+        FROM scratch
+        MAINTAINER docker
+        ENV foo bar`
+
+	head, out, err := buildImageWithOut("scratch-image", dockerfile, false)
+	c.Assert(err, check.IsNil)
+
+	split := strings.Split(out, "\n")
+	intermediate := strings.TrimSpace(split[5][7:])
```
Comment (aaronlehmann) [inline integration-cli/docker_cli_images_test.go]: Why the numbers 5 and 7? This probably deserves a comment (or parsing that's less sensitive to the output format).

File: graph/graph.go (hunk 4 above)
Comment (tonistiigi) [inline graph/graph.go]: so you want to check that `img.ID != img.Parent` ?

File: integration-cli/docker_cli_images_test.go [hunk 6]
```diff
@@ -197,3 +196,42 @@ func (s *DockerSuite) TestImagesEnsureDanglingImageOnlyListedOnce(c *check.C) {
...
+
+	out, _ = dockerCmd(c, "images")
+	if strings.Contains(out, intermediate) {
+		c.Fatalf("images shouldn't show non-heads images, got %s in %s", intermediate, out)
+	}
+	if !strings.Contains(out, head[:12]) {
```
Comment (aaronlehmann) [inline integration-cli/docker_cli_images_test.go]: Use `stringid.shortLen` instead of 12.

File: integration-cli/docker_cli_images_test.go [hunk 7]
```diff
@@ -197,3 +196,42 @@ func (s *DockerSuite) TestImagesEnsureDanglingImageOnlyListedOnce(c *check.C) {
...
+
+	dockerfile := `
+        FROM scratch
+        MAINTAINER docker`
+
+	id, _, err := buildImageWithOut("scratch-image", dockerfile, false)
+	c.Assert(err, check.IsNil)
+
+	out, _ := dockerCmd(c, "images")
+	if !strings.Contains(out, id[:12]) {
```
Comment (aaronlehmann) [inline integration-cli/docker_cli_images_test.go]: Same here - use `stringid.shortLen`.

File: integration-cli/docker_cli_images_test.go (hunk 6 above)
Comment (runcom) [inline integration-cli/docker_cli_images_test.go]: it's not exported and lots of code deal with `:12`, probably will need a whole refactor if we decide to export shortlen from stringid

File: graph/graph.go (hunk 3 above)
Comment (runcom) [inline graph/graph.go]: done

File: integration-cli/docker_cli_images_test.go (hunk 5 above)
Comment (runcom) [inline integration-cli/docker_cli_images_test.go]: updated with comment
need intermediate img id, i don't know any other way to get one here

File: integration-cli/docker_cli_images_test.go (hunk 6 above)
Comment (runcom) [inline integration-cli/docker_cli_images_test.go]: I'll make a follow PR to fix this in the whole project (this has enough commits)

Comment (LK4D4) [inline integration-cli/docker_cli_images_test.go]: I agree, this PR already contains unrelated changes.

File: graph/graph.go [hunk 8]
```diff
@@ -174,6 +179,20 @@ func (graph *Graph) restore() error {
 	for _, v := range dir {
 		id := v.Name()
 		if graph.driver.Exists(id) {
+			pth := filepath.Join(graph.root, id, "json")
```
Comment (tonistiigi) [inline graph/graph.go]: Why not `graph.loadImage(id)`? After #16953 the parent reference may not be in the raw json file. "json" is a constant `jsonFileName`.

File: graph/graph.go (hunk 3 above)
Comment (tiborvass) [inline graph/graph.go]: @runcom I don't see the comment :S

Comment (runcom) [inline graph/graph.go]: hahah

File: graph/graph.go (hunk 8 above)
Comment (runcom) [inline graph/graph.go]: fair, i'll change it

File: graph/graph.go [hunk 9]
```diff
@@ -174,6 +179,20 @@ func (graph *Graph) restore() error {
...
+			jsonSource, err := os.Open(pth)
+			if err != nil {
+				return err
+			}
+			defer jsonSource.Close()
+			decoder := json.NewDecoder(jsonSource)
+			img := &image.Image{}
```
Comment (tiborvass) [inline graph/graph.go]: you could leave `img` on the stack by removing `&` and doing `Decode(&img)`

**Comment by aaronlehmann** [comment]
Now that https://github.com/docker/docker/pull/16953 has been merged, this needs to have some kind of locking to prevent `register` and `attemptIDReuse` from running at the same time. They both lock multiple images at once, but potentially obtain the locks in different orders, so there is a potential deadlock.

Aside from that, this looks good to me.

**Comment by runcom** [comment]
@aaronlehmann can we lock the whole graph struct since `V2Puller` holds a reference to it when locking? maybe use this additional lock just for this synchronization. Any other ideas?

**Comment by aaronlehmann** [comment]
Currently, there is a lock called `idReuseLock`:

https://github.com/docker/docker/blob/master/graph/pull_v2.go#L552

This could be renamed to something generic and moved to the `Graph` struct, to make a lock used by anything that locks more than one image mutex at a time.

But I'm wondering if it's really necessary for `register` to lock two mutexes at once. Can't it call `graph.imageMutex.Unlock(imgID)` before calling `graph.imageMutex.Lock(parent)`? That would work around the issue.

**Comment by runcom** [comment]
> But I'm wondering if it's really necessary for register to lock two mutexes at once. Can't it call graph.imageMutex.Unlock(imgID) before calling graph.imageMutex.Lock(parent)? That would work around the issue.

This is much better, and there shouldn't be any issue in doing that

**Comment by runcom** [comment]
also, it's not `register` that locks, but the exported `Register`, I'll see what I can do

**Comment by thaJeztah** [comment]
ping @aaronlehmann I see this PR was updated, PTAL

**Comment by aaronlehmann** [comment]
@thaJeztah: My last comment hasn't been addressed yet, as far as I can tell.

**Comment by thaJeztah** [comment]
@aaronlehmann np, I thought you might've missed the update (didn't look in the code changes)

**Comment by runcom** [comment]
@aaronlehmann went with avoiding locking twice in `register` instead of adding a lock to the graph. PTAL

**Comment by aaronlehmann** [comment]
@runcom Graph unit tests appear to be hanging.

**Comment by tonistiigi** [comment]
@runcom Note that some code paths like v2 pulls call `graph.register()` directly with the pull code itself holding the lock. Currently with this patch all v2 pulls deadlock because of this.

**Comment by runcom** [comment]
Right, I'll fix that

**Comment by runcom** [comment]
rebased and updated

**Comment by cpuguy83** [comment]
vet: integration-cli/docker_cli_rmi_test.go: integration-cli/docker_cli_rmi_test.go:235:1: expected statement, found '<<' (and 2 more errors)

**Comment by runcom** [comment]
ouch rebase, fixed

**Comment by runcom** [comment]
`TestConcurrentPullMultipleTags` seems to hang in experimental, will look into this

**Comment by cpuguy83** [comment]
So this worked well for me, tried pulling while doing rmi didn't see any deadlocks.
That said, I'm not confident enough to give an LGTM in this part of the codebase.

**Comment by tonistiigi** [comment]
LGTM

**Comment by runcom** [comment]
finally green :) ping @tiborvass @aaronlehmann

**Comment by tiborvass** [comment]
@runcom what did you change in the last push?

**Comment by runcom** [comment]
@tiborvass just rebased

**Comment by tiborvass** [comment]
I tested it, it's much faster. weak lgtm for the code

**Comment by cpuguy83** [comment]
ping @aaronlehmann

**Comment by tiborvass** [comment]
YOLO

**Comment by tiborvass** [comment]
Removing impact/changelog because it seems this performance issue was introduced after 1.8. @runcom let me know if i'm wrong.

**Comment by tiborvass** [comment]
Apparently i'm wrong :) adding it back
